docker/build_support/build_image

```diff
@@ -0,0 +1,118 @@
+#! /usr/bin/env bash
+set -e
+
+prog=$0
+error() {
+   echo "Whoops!  Looks like $0:$2 failed."
+   echo "Please try rerunning $prog again."
+   exit 1
+}
+trap 'error "${BASH_SOURCE}" "${LINENO}"' ERR
+
+SOURCE="${BASH_SOURCE[0]}"
+while [ -h "$SOURCE" ]; do
+  DIR="$( cd -P "$( dirname "$SOURCE" )" && pwd )"
+  SOURCE="$(readlink "$SOURCE")"
+  [[ $SOURCE != /* ]] && SOURCE="$DIR/$SOURCE" 
+done
+DIR="$( cd -P "$( dirname "$SOURCE" )" && pwd )"
+
+function usage() {
+  echo "$0 <Dockerfile> [<dep> [<dep2> ...]]" 1>&2
+	cat $SOURCE | grep -E '^#-.*$$' | sed -e 's@^#- @  @' 1>&2
+}
+
+RED='\033[0;31m'
+YELLOW='\033[1;33m'
+CYAN='\033[1;36m'
+NC='\033[0m'
+
+if [[ "$1" == "-h" ]]; then
+  usage
+  exit 1
+fi
+
+#- Builds the given Dockerfile into an image,
+#-   first by constructing a build directory that links together the shortest
+#-   common parent directories of given dependencies, then running the given Dockerfile
+#-   in that context.  It also injects metadata related to the release sha at the time
+#-   the image is built.
+Dockerfile=$(readlink -m $1)
+
+baseTag=$(basename "$(dirname "$(realpath $Dockerfile)")")
+
+#- The final tag will also include env variables IMAGES_PREFIX and IMAGES_POSTFIX
+tag="${IMAGES_PREFIX}${baseTag}${IMAGES_POSTFIX}"
+
+function imageExists() {
+  docker inspect --type=image $tag &>/dev/null
+}
+
+if [[ -z "$Dockerfile" ]]; then
+    usage
+    exit 1
+fi
+
+shift
+
+buildDir="$(mktemp -d)"
+sourceDockerDir="$(readlink -m $(dirname $Dockerfile))"
+rsync -a $sourceDockerDir/ $buildDir/
```

Comment:
So basically, first copy the source code into a 'build' directory (temp directory) that we can manipulate a bit.